pkg/client/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,219 @@
+	OnStartedLeading func(stop <-chan struct{})
+	OnStoppedLeading func()
+}
+
+// possible future async callbacks:
+//  * OnChallenge()
+//  * OnNewLeader()
+// possible future synchronous calls:
+//  * IsLeader()
+//  * GetLeader()
+type LeaderElector struct {
+	config LeaderElectionConfig
+	// internal bookkeeping
+	observedRecord LeaderElectionRecord
+	observedTime   time.Time
+}
+
+// TODO(mikedanese): this should potentially be versioned
+type LeaderElectionRecord struct {
+	HolderIdentity       string           `json:"holderIdentity"`
+	LeaseDurationSeconds int              `json:"leaseDurationSeconds"`
+	AquireTime           unversioned.Time `json:"aquireTime"`
+	RenewTime            unversioned.Time `json:"renewTime"`
+}
+
+// Run starts the leader election loop
+func (le *LeaderElector) Run() {
+	defer func() {
+		util.HandleCrash()
+		le.config.Callbacks.OnStoppedLeading()
+	}()
+	le.aquire()
+	stop := make(chan struct{}, 1)
+	go le.config.Callbacks.OnStartedLeading(stop)
+	le.renew()
+	stop <- struct{}{}
+}
+
+// aquire loops calling tryAquireOrRenew and returns immediately when tryAquireOrRenew succeeds.
+func (le *LeaderElector) aquire() {
+	stop := make(chan struct{}, 1)
+	util.Until(func() {
+		succeeded := le.tryAquireOrRenew()
+		if !succeeded {
+			glog.V(4).Infof("failed to renew lease %v/%v", le.config.EndpointsMeta.Namespace, le.config.EndpointsMeta.Name)
+			time.Sleep(wait.Jitter(le.config.RetryPeriod, JitterFactor))
+			return
+		}
+		le.config.EventRecorder.Eventf(&api.Endpoints{ObjectMeta: le.config.EndpointsMeta}, "%v became leader", le.config.Identity)
+		glog.Infof("sucessfully aquired lease %v/%v", le.config.EndpointsMeta.Namespace, le.config.EndpointsMeta.Name)
+		stop <- struct{}{}
+	}, 0, stop)
+}
+
+// renew loops calling tryAquireOrRenew and returns immediately when tryAquireOrRenew fails.
+func (le *LeaderElector) renew() {
+	stop := make(chan struct{}, 1)
+	util.Until(func() {
+		err := wait.Poll(le.config.RetryPeriod, le.config.RenewDeadline, func() (bool, error) {
+			return le.tryAquireOrRenew(), nil
+		})
+		if err == nil {
+			glog.V(4).Infof("succesfully renewed lease %v/%v", le.config.EndpointsMeta.Namespace, le.config.EndpointsMeta.Name)
+			return
+		}
+		le.config.EventRecorder.Eventf(&api.Endpoints{ObjectMeta: le.config.EndpointsMeta}, "%v stopped leading", le.config.Identity)
+		glog.Infof("failed to renew lease %v/%v", le.config.EndpointsMeta.Namespace, le.config.EndpointsMeta.Name)
+		stop <- struct{}{}
+	}, 0, stop)
+}
+
+// tryAquireOrRenew tries to aquire a leader lease if it is not already aquired,
+// else it tries to renew the lease if it has already been aquired. Returns true
+// on success else returns false.
+func (le *LeaderElector) tryAquireOrRenew() bool {
+	now := unversioned.Now()
+	leaderElectionRecord := LeaderElectionRecord{
+		HolderIdentity:       le.config.Identity,
+		LeaseDurationSeconds: int(le.config.LeaseDuration / time.Second),
+		RenewTime:            now,
```

Comment:
You are marshalling time.Now() using the default formatter, which I think prints out the time in local time zone.  But nodes may be in different time zones.  Are two nodes guaranteed to be able to parse time zones used by other zones?  Seems like it depends on ZONEINFO and such.  I know nodes in a cluster are supposed to have synchronized clocks, but I don't know what syncs time zone info.  Suggest either format time in UTC or document this requirement.